### PR TITLE
Expose JSON encoding errors

### PR DESCRIPTION
**Pull request by tomtoday**
JSON encoding errors in code like `generate_json.php` would fail
silently (resulting in JSON output `{“status”: “404”, “error”:
“Resource Not Found”}`), this adds the ability to return a 500 error
with a better error message. It also handles bad UTF-8 encoding issues
rather than erroring.

### CODE REVIEW

**Comment by tomtoday** [comment]
@mryand and @laupow Can you guys please take a look at this? Let me know if you have any questions.

**Comment by laupow** [comment]
Exposing the error condition in the JSON response is a real improvement.

I wonder what conditions cause json_encode to fail with JSON_ERROR_UTF8. It could be related to some of the database charset quirks again. No fun is to be had digging there.

**Comment by tomtoday** [comment]
I saw the issue with a fresh vanilla install with docker (
https://github.com/reasoncms/reasoncms/blob/4.8/docker/README.md). I
suspect it has to do with collation on mariadb itself since that has not
been altered in that situation. The DB dump (reason4.8.sql) itself looks to
be correct for utf-8

File: reason_4.0/lib/core/classes/api/feed/feed.php [hunk 1]
```diff
@@ -110,7 +110,11 @@ class ReasonFeedAPI extends CarlUtilAPI
 				if ($json = $model->get())
 				{
 					$this->set_content($json);
-				}	
+					// pull out the status and set it on the response
+					$decoded_json = json_decode($json, true);
```
Comment (mryand) [inline reason_4.0/lib/core/classes/api/feed/feed.php]: An alternate approach would be to add an additional method to the model, like `get_response_code()`, which would remove the need to parse the output and reduce the likelihood of a namespace collision (e.g. some json format that uses a "status" key).

Then the code on this side would look something like:
```php
$this->set_http_response_code($model->get_response_code());
```

**Comment by tomtoday** [comment]
That's a good call @mryand. I would like to bring all of the JSON modesl in line with this but it requires a refactoring of `ReasonImageJSON` to use `ReasonMVCModel`, I might take a look at that.

**Comment by mryand** [comment]
That looks good. Since we are adding methods to `ReasonFeedInterface` we should probably double check that all implementing classes use those methods (if you haven't already).

**Pull request merged**